### PR TITLE
new Pareto-k diagnostics vignette

### LINKED ISSUE #322: Vignette for Pareto diagnostics
I made a notebook about the Pareto diagnostics https://users.aalto.fi/~ave/casestudies/Pareto/pareto_diagnostics.html
I can turn this to a vignette, but looking for a feedback first

### PR DESCRIPTION
**Pull request by avehtari**
Fixes #322

### CODE REVIEW

**Comment by paul-buerkner** [comment]
Thanks! Do you think this vignette needs additional review? If yes, could @n-kall do it since you have been implementing these methods.

**Comment by n-kall** [comment]
Sure, I can look at this

File: vignettes/pareto_diagnostics.Rmd [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+                   xt1_5=qt(pnorm(xn), df=1.5),
+                   xt1=qt(pnorm(xn), df=1))
+```
+
+### MCMC convergence diagnostics
+
+We examine the draws with the default `summarise_draws()`.
+
+```{r summarise_draws}
+drt |> summarise_draws()
```
[the remark this answers is not in the document]
Comment (n-kall) [reply vignettes/pareto_diagnostics.Rmd:71]: Would it be better here to use `default_convergence_measures()`to focus on the convergence diagnostics at first?

File: vignettes/pareto_diagnostics.Rmd [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+                   xt1_s=pareto_smooth(xt1)) |>
+  subset_draws(variable="_s", regex=TRUE)
+```
+
+Now the `mcse_mean` values are more trustworthy when $\hat{k} < 0.7$.
+When $\hat{k}>0.7$ both bias and variance grow so fast that Pareto smoothing
+rarely helps (see more details in the paper).
+
+```{r summarise_draws-pareto_khat-2}
+drts |> summarise_draws(mean, mcse_mean, ess_basic, pareto_khat)
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:145]:
```suggestion
summarise_draws(drts, mean, mcse_mean, ess_basic, pareto_khat)
```

File: vignettes/pareto_diagnostics.Rmd [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+drt |> summarise_draws(mean, mcse_mean, ess_basic,
+                       pareto_khat, min_ss=pareto_min_ss)
+```
+
+Here required `min_ss` is smaller than `ess_basic` for all except $t_1$, for
+which there is no hope.
+
+## Convergence rate
+
+Given finite variance, central limit theorem states that to halve
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:165]:
```suggestion
Given finite variance, the central limit theorem states that to halve
```

File: vignettes/pareto_diagnostics.Rmd [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+MCSE we need four times bigger sample size. With Pareto smoothing,
+we can go further, but the convergence rate decreases when $\hat{k}$ increases.
+
+```{r summarise_draws-conv_rate}
+drt |> summarise_draws(mean, mcse_mean, ess_basic,
+                       pareto_khat, min_ss=pareto_min_ss,
+                       conv_rate=pareto_convergence_rate)
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:172]:
```suggestion
summarise_draws(drt, mean, mcse_mean, ess_basic,
                       pareto_khat, min_ss = pareto_min_ss,
                       conv_rate = pareto_convergence_rate)
```

File: vignettes/pareto_diagnostics.Rmd [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+## Pareto-$\hat{k}$-threshold
+
+The final Pareto diagnostic, $\hat{k}$-threshold, is useful for smaller sample sizes. Here we select only 100 iterations per chain to get total of 400 draws.
+
+```{r summarise_draws-khat_threshold}
+drt |> subset_draws(iteration=1:100) |>
+  summarise_draws(mean, mcse_mean, ess_basic,
+                  pareto_khat, min_ss=pareto_min_ss,
+                  khat_thres=pareto_khat_threshold,
+                  conv_rate=pareto_convergence_rate)
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:188]:
```suggestion
drts <- subset_draws(drt, iteration=1:100)

drts <- summarise_draws(drts, mean, mcse_mean, ess_basic,
                  pareto_khat, min_ss = pareto_min_ss,
                  khat_thres = pareto_khat_threshold,
                  conv_rate = pareto_convergence_rate)
```

File: vignettes/pareto_diagnostics.Rmd [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+## Pareto diagnostics
+
+We can get all these diagnostics with `pareto_diags()`, and it's
+easy to use it also for derived quantities.
+
+```{r summarise_draws-pareto_diags}
+drt |> mutate_variables(xt2_5_sq=xt2_5^2) |> subset_draws(variable="xt2_5_sq") |>
+  summarise_draws(mean, mcse_mean,
+                  pareto_diags)
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:203]:
```suggestion
drtm <- mutate_variables(drt, xt2_5_sq = xt2_5^2)
drtm <- subset_draws(drtm, variable = "xt2_5_sq")
drtm <- summarise_draws(drtm, mean, mcse_mean, pareto_diags)
```

File: vignettes/pareto_diagnostics.Rmd [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+```
+
+### Simulated data
+
+Generate `xn` a simulated MCMC sample with 4 chains each with 1000
+iterations using AR process with marginal normal(0,1)
+
+```{r simulate-data-1}
+N=1000
+phi=0.3
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:41]:
```suggestion
N <- 1000
phi <- 0.3
```

File: vignettes/pareto_diagnostics.Rmd [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    toc_depth: 3
+vignette: >
+  %\VignetteIndexEntry{Pareto-khat diagnostics}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+## Introduction 
+
+Paper
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:16]:
```suggestion
The paper
```

File: vignettes/pareto_diagnostics.Rmd [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+
+We examine the draws with the default `summarise_draws()`.
+
+```{r summarise_draws}
+drt |> summarise_draws()
+```
+
+All the usual convergence diagnostics $\widehat{R}$, Bulk-ESS, and
+Tail-ESS look good, which is fine as they have been designed to
+work also with infinite variance [@Vehtari+etal:2021:Rhat].
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:76]: Is the Rhat reference defined somewhere?

File: vignettes/pareto_diagnostics.Rmd [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+we can go further, but the convergence rate decreases when $\hat{k}$ increases.
+
+```{r summarise_draws-conv_rate}
+drt |> summarise_draws(mean, mcse_mean, ess_basic,
+                       pareto_khat, min_ss=pareto_min_ss,
+                       conv_rate=pareto_convergence_rate)
+```
+
+We see that with $t_2$, $t_{1.5}$, and $t_1$  we need $4^{1/0.86}\approx 5$,
+$4^{1/0.60}\approx 10$, and $4^{1/0}\approx \infty$ bigger sample sizes to
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:176]:
```suggestion
$4^{1/0.60}\approx 10$, and $4^{1/0}\approx \infty$ times bigger sample sizes to
```

File: vignettes/pareto_diagnostics.Rmd [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+* Aki Vehtari, Daniel Simpson, Andrew Gelman, Yuling Yao, and
+  Jonah Gabry (2024). Pareto smoothed importance sampling.
+  *Journal of Machine Learning Research*, 25(72):1-58.
+
+presents Pareto smoothed importance sampling, but also
+Pareto-$\hat{k}$ diagnostic that can be used when estimating any
+expectation based on finite sample. This vignette illustrates the use of
+these diagnostics.
+
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:26]: Maybe it makes sense to list the functions at the beginning. Something like this:

```suggestion
The individual diagnostic functions are `pareto_khat()`, `pareto_min_ss()`, `pareto_convergence_rate()` and `pareto_khat_threshold(). The function `pareto_diags()` will return all of these.

Additionally, the `pareto_smooth()` function can be used to transform draws by smoothing the tail(s).
```

**Review by n-kall — COMMENTED** [review]
I haven't yet had a chance to run the vignette, but as the tests were passing (except on R 3.6) I have gone through and made some suggestions.

Most suggestions are removing the `|>` operator which is not available in R < 4.1.0. If we want to use the native pipe, I think would need to bump the required version to 4.1.0 (as far as I understand)

File: vignettes/pareto_diagnostics.Rmd [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+```{r simulate-data-1}
+N=1000
+phi=0.3
+set.seed(6534)
+dr<-array(data=replicate(4,as.numeric(arima.sim(n = N,
+                                                list(ar = c(phi)),
+                                                sd = sqrt((1-phi^2))))),
+         dim=c(N,4,1)) |>
+  as_draws_df() |>
+  set_variables('xn')
```
[the remark this answers is not in the document]
Comment (avehtari) [reply vignettes/pareto_diagnostics.Rmd:48]: As `posterior` suggests `dplyr`, can I use `%>%`?

[the remark this answers is not in the document]
Comment (paul-buerkner) [reply vignettes/pareto_diagnostics.Rmd:48]: Yes, I think so

[the remark this answers is not in the document]
Comment (n-kall) [reply vignettes/pareto_diagnostics.Rmd:50]: Good point. dplyr is importing magrittr so should work. That will be a much better solution

File: vignettes/pareto_diagnostics.Rmd (hunk 1 above)
[the remark this answers is not in the document]
Comment (avehtari) [reply vignettes/pareto_diagnostics.Rmd]: I like it as it is now

File: vignettes/pareto_diagnostics.Rmd (hunk 12 above)
[the remark this answers is not in the document]
Comment (avehtari) [reply vignettes/pareto_diagnostics.Rmd]: Switched to `%>%`

File: vignettes/pareto_diagnostics.Rmd (hunk 9 above)
Comment (avehtari) [reply vignettes/pareto_diagnostics.Rmd]: Added

**Comment by avehtari** [comment]
Thanks Noa. 

I made the suggested, except for using `%>%` and not using `default_convergence_measures()`. 

I also fixed the two references in `posterior.Rmd`. Could have been a separate from this PR, but maybe this is ok

File: vignettes/pareto_diagnostics.Rmd [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+
+* Aki Vehtari, Daniel Simpson, Andrew Gelman, Yuling Yao, and
+  Jonah Gabry (2024). Pareto smoothed importance sampling.
+  *Journal of Machine Learning Research*, 25(72):1-58.
+
+presents Pareto smoothed importance sampling, but also
+Pareto-$\hat{k}$ diagnostic that can be used when estimating any
+expectation based on finite sample. This vignette illustrates the use of
+these diagnostics.
+The individual diagnostic functions are `pareto_khat()`, `pareto_min_ss()`, `pareto_convergence_rate()` and `pareto_khat_threshold(). The function `pareto_diags()` will return all of these.
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:26]: Missed `

```suggestion
The individual diagnostic functions are `pareto_khat()`, `pareto_min_ss()`, `pareto_convergence_rate()` and `pareto_khat_threshold()`. The function `pareto_diags()` will return all of these.
```

File: vignettes/pareto_diagnostics.Rmd [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+We examine the draws with the default `summarise_draws()`.
+
+```{r summarise_draws}
+drt %>%
+  summarise_draws()
+```
+
+All the usual convergence diagnostics $\widehat{R}$, Bulk-ESS, and
+Tail-ESS look good, which is fine as they have been designed to
+work also with infinite variance (Vehtari et al., 2022).
```
Comment (n-kall) [inline vignettes/pareto_diagnostics.Rmd:80]: I think the year should be 2020
```suggestion
work also with infinite variance (Vehtari et al., 2020).
```

File: vignettes/posterior.Rmd [hunk 15]
```diff
@@ -396,11 +396,11 @@ with `draws` objects, please open an issue at
...
 
-Gelman A., Carlin J. B., Stern H. S., David B. Dunson D. B., Aki Vehtari A.,
+Gelman A., Carlin J. B., Stern H. S., David B. Dunson D. B., Vehtari, A.,
 & Rubin D. B. (2013). *Bayesian Data Analysis, Third Edition*. Chapman and
 Hall/CRC.
 
 Vehtari A., Gelman A., Simpson D., Carpenter B., & Bürkner P. C. (2020).
 Rank-normalization, folding, and localization: An improved Rhat for assessing
-convergence of MCMC. *Bayesian Analysis*.
+convergence of MCMC. *Bayesian Analysis*, , 16(2):667-718.
```
Comment (n-kall) [inline vignettes/posterior.Rmd:405]: Accidental additional comma?

```suggestion
convergence of MCMC. *Bayesian Analysis*, 16(2):667-718.
```

**Comment by avehtari** [comment]
Thanks Noa, you have magical eyes for spotting errors!

**Comment by n-kall** [comment]
Looks good! All the vignettes are failing to build with R 3.6 but I think that is due to the `knitr` / `evaluate`. 
I think this could be merged

**Pull request merged**